From 63f7d32eed37e56c5de7b0140b81f70388d15d69 Mon Sep 17 00:00:00 2001
From: Karthikeyan Kathirvel <quic_kathirve@quicinc.com>
Date: Tue, 13 Jun 2023 11:43:25 +0530
Subject: [PATCH] ath12k: drop the non transmitted failure tx frames

In mesh node traffic, internal fw tx failures are reported as a
transmitted failure to mesh metric calculation and hence mesh link is
broken.

Fix the issue by dropping the internal fw tx failures at driver which
prevents false failure averaging of mesh metric calculation.

Signed-off-by: Karthikeyan Kathirvel <quic_kathirve@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/dp_tx.c b/drivers/net/wireless/ath/ath12k/dp_tx.c
index 52b24fe70eea..1d153476a276 100644
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -1140,6 +1140,18 @@ static void ath12k_dp_tx_complete_msdu(struct ath12k *ar,
 	status.rates = &status_rate;
 	status.n_rates = 1;
 
+	if (ts.status != HAL_WBM_TQM_REL_REASON_FRAME_ACKED) {
+		switch (ts.status) {
+		case HAL_WBM_TQM_REL_REASON_CMD_REMOVE_MPDU:
+		case HAL_WBM_TQM_REL_REASON_DROP_THRESHOLD:
+		case HAL_WBM_TQM_REL_REASON_CMD_REMOVE_AGED_FRAMES:
+			dev_kfree_skb_any(msdu);
+			return;
+		default:
+			//TODO: Remove this print and add as a stats
+			ath12k_dbg(ab, ATH12K_DBG_DP_TX, "tx frame is not acked status %d\n", ts.status);
+		}
+	}
 
 	if (unlikely(ath12k_debugfs_is_extd_tx_stats_enabled(ar))) {
 		if(arsta->wbm_tx_stats && wbm_status < HAL_WBM_REL_HTT_TX_COMP_STATUS_MAX)
-- 
2.38.0

